controllers/contentlibrary/clustercontentlibraryitem/clustercontentlibraryitem_controller.go

```diff
@@ -159,9 +155,12 @@ func (r *Reconciler) ReconcileNormal(ctx goctx.Context, cclItem *imgregv1a1.Clus
 			savedStatus = cvmi.Status.DeepCopy()
 		}()
 
-		if utils.CheckItemReadyCondition(cclItem.Status.Conditions) {
-			conditions.MarkTrue(cvmi, vmopv1.VirtualMachineImageProviderReadyCondition)
-		} else {
+		if !controllerutil.ContainsFinalizer(cvmi, utils.ClusterVirtualMachineImageFinalizer) {
```

Comment:
This doesn't look correct to me: this controller watches the ClusterContentLibraryItem, and that's what needs our finalizer if we need to ensure that the ClusterContentLibraryItem is not removed before we do our cleanup. Here, by adding the finalizer on the ClusterVirtualMachineImage, the CCLI Owner could go away but the image is orphaned b/c the finalizer is still present.

Reply:
`ReconcileDelete()` is only guaranteed to be reached if the `Get()` in `Reconcile()` succeeds. A controller can ensure that succeeds - and is able to observe `!DeletionTimestamp.IsZero()` - by adding its own finalizer on the object being watched. Here, yes, the OwnerRef would cause GC to set DeletionTimestamp on the image when the item is marked for delete, but then you're racing with the image registry controller to observe the delete timestamp before it updates the item, removing its finalizer, and then causing the item removal from etcd. You can experiment with this behavior by creating a parent/child CMs and using `kubectl delete/edit` to play the role of controller (the OwnerRef BlockOwnerDeletion only applies to foreground deletes).

> To address this, I updated the PR to set the controller reference first before adding the finalizer

I don't get why this order would matter. Both fields are a part of the ObjectMeta, and would be sent to the api-server together regardless.

Reply:
Ah, I didn't realize that K8s API uses background cascading deletion by default. What you said above totally makes sense now and I have updated the PR to add the finalizer in cl/ccl-item resources as suggested.

Regarding the image deletion case I mentioned earlier (where the item itself is not deleted), I figured it should be fine since the controller will recreate the image once it reconciles the item resource.

> I don't get why this order would matter.

In the previous implementation, the owner reference wouldn't be set if the item condition wasn't ready while leaving the image with a finalizer. This should no longer be a problem as the finalizer is added to item resources now. However, I still kept this change so that we always set up the image spec/status based on existing item resources, and update the image condition afterwards.